functions.php

```diff
@@ -0,0 +1,48 @@
+<?php
+    session_start();
+    //render the csrf token input field
+    function csrfToken()
+    {
+        if(function_exists("random_bytes"))
+        {
+            $token = $_SESSION['token'] = bin2hex(random_bytes(20));
+        }
+        else
+        {
+            $token = $_SESSION['token'] = bin2hex(openssl_random_pseudo_bytes(20));
+        }
+        $field = "<input type='hidden' name='token' value=$token>";
+        echo $field;
+    }
+    //check if the csrf token is valid
+    function validateToken()
+    {
+        if(!hash_equals($_SESSION['token'],$_POST['token']))
+        {
+            die("Invalid Token!");
+        }
+    }
+    //check if a parameter is indeed numeric
+    function checkNumericParam($param)
+    {
+        if(!is_numeric($param))
+        {
+            die("Invalid Request!");
```

Comment:
You're saying that the message is unclear or that the die() function shouldn't be used?

Reply:
I think that a better error message would be enough. But since you mention it, `die()` kills the PHP script at this line, so if you have done something up to this point, it has already been executed. For example, you may add something to the database, then call something that can `die()`, and then you may delete it. A lot of developers usually forget that the deletion will never execute. I am not saying that it's happening now, just something to keep in mind in general.

Reply:
Yup the error message should be improved indeed.<br> I'll keep that in mind regarding the die() function.